universe/Building.h

```diff
@@ -214,9 +214,8 @@ class BuildingTypeManager {
 
 private:
     BuildingTypeManager();
-    ~BuildingTypeManager();
 
-    std::map<std::string, BuildingType*> m_building_types;
+    std::map<std::string, std::unique_ptr<BuildingType>> m_building_types;
```

Comment:
tl;dr; Yes, that is not thread safe and can crash.

It is unsafe at the hardware level.  It is described in the [Intel Guide for Developing Multithreaded Applications](https://software.intel.com/en-us/articles/intel-guide-for-developing-multithreaded-applicationsl).  

At the hardware level, a read is not a passive observation.  In the middle of the read the data at the storage location may not exist and will be re-written at the end of the read.  If two threads try to read the same location without synchronization, undefined behavior happens.  

I will try and clarify what I wrote 3 posts ago,
> "shared_ptr and weak_ptr are as thread safe as a raw pointer."  

The smart pointers are designed to be as safe as a raw pointer, but not more.

For example, if you have two threads with a raw pointer in each respective thread pointing to a common object. 

Copying the pointers or reading the address from the pointers with no consideration for thread synchronization is safe because each pointer is in memory only accessed by their respective thread, so those operations don't interact with any shared memory.  

Dereferencing the pointers (any read or write of the pointed to object) without consideration for thread synchronization causes undefined behavior.  This is true even for primitive types, POD, const function or data members, or dynamic casts.

Operations on the pointers themselves are safe.  Any operations on the pointed to object are unsafe.

The std smart pointers, including the `shared_ptr` have the same guarantee.  Operations on the pointers themselves, all of the member functions constructing, copying, moving, etc are safe.  The smart pointers use synchronization techniques internally to guarantee that.  Any operations on the pointed too object is unsafe.

The statement you referenced is intended to clarify that operations on the pointers themselves, specifically copying which increments counters in memory shared between the threads is safe in multiple threads.